crates/holochain_persistence_lmdb/src/eav/lmdb.rs

```diff
@@ -0,0 +1,299 @@
+
+        let env = manager
+            .read()
+            .expect("Could not get a read lock on the manager");
+
+        // Then you can use the environment handle to get a handle to a datastore:
+        let options = StoreOptions {
+            create: true,
+            flags: DatabaseFlags::empty(),
+        };
+        let store: SingleStore = env
+            .open_single(EAV_BUCKET, options)
+            .expect("Could not create EAV store");
+
+        EavLmdbStorage {
+            id: Uuid::new_v4(),
+            store: store,
+            manager: manager.clone(),
+            attribute: PhantomData,
+        }
+    }
+}
+
+impl<A: Attribute> Debug for EavLmdbStorage<A> {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        f.debug_struct("EavLmdbStorage")
+            .field("id", &self.id)
+            .finish()
+    }
+}
+
+impl<A: Attribute> EavLmdbStorage<A>
+where
+    A: Sync + Send + serde::de::DeserializeOwned,
+{
+    fn add_lmdb_eavi(
+        &mut self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> Result<Option<EntityAttributeValueIndex<A>>, StoreError> {
+        let env = self.manager.read().unwrap();
+        let mut writer = env.write()?;
+
+        // use a clever key naming scheme to speed up exact match queries on the entity
+        let key = format!("{}::{}", eav.entity(), eav.index());
+
+        self.store
+            .put(&mut writer, key, &Value::Json(&eav.content().to_string()))?;
+
+        writer.commit()?;
+
+        Ok(Some(eav.clone()))
+    }
+
+    fn fetch_lmdb_eavi(
+        &self,
+        query: &EaviQuery<A>,
+    ) -> Result<BTreeSet<EntityAttributeValueIndex<A>>, StoreError> {
+        let env = self.manager.read().unwrap();
+        let reader = env.read()?;
+
+        let entries = match &query.entity {
+            EavFilter::Exact(entity) => {
+                // Can optimize here thanks to the sorted keys and only iterate matching entities
+                self.store
+                    .iter_from(&reader, format!("{}::{}", entity, 0))? // start at the first key containing the entity address
+                    .filter_map(Result::ok)
+                    .take_while(|(k, _v)| {
+                        // stop at the first key that doesn't match
+                        String::from_utf8(k.to_vec())
+                            .unwrap()
+                            .contains(&entity.to_string())
+                    })
+                    .filter_map(|(_k, v)| match v {
+                        Some(Value::Json(s)) => serde_json::from_str(&s).ok(),
+                        _ => None,
+                    })
+                    .collect::<BTreeSet<EntityAttributeValueIndex<A>>>()
+            }
+
+            _ => {
+                // In this case all we can do is iterate the entire database
+                self.store
+                    .iter_start(&reader)?
+                    .filter_map(Result::ok)
```

Comment:
What are the error items you're throwing away here? Just curious.

Reply:
I assume any database read errors and later on in the combinator it throws away conversion errors.

I'm not quite sure what the best way to handler errors here is. The other extreme is one error causes the whole thing to error maybe that is better

Reply:
Ok I rewrote this to bail on the first error it encounters and return that